perl-package/AI-MXNet/lib/AI/MXNet/KVStore.pm

```diff
@@ -28,35 +28,73 @@ use AI::MXNet::Function::Parameters;
 
 =head1 NAME
 
-    AI::MXNet::KVStore - Key value store interface of MXNet.
+    AI::MXNet::KVStore - Key value store interface of MXNet for parameter synchronization.
+=cut
+func _ctype_key_value($keys, $vals)
+{
+    if(ref $keys eq 'ARRAY') {
+        assert(@$keys == @$vals);
+        my @c_keys;
+        my @c_vals;
+        zip(sub {
+            my($key, $val) = @_;
+            my($c_key_i, $c_val_i) = _ctype_key_value($key, $val);
+            push @c_keys, @$c_key_i;
+            push @c_vals, @$c_key_i;
+        }, $keys, $vals);
+        return (\@c_keys, \@c_vals);
+    }
+    my @names;
+    if(blessed $vals and $vals->isa('AI::MXNet::NDArray')) {
+        push @names, "" . $keys;
+        return (\@names, [ $vals->handle ]);
+    } else {
+        assert(blessed $_ and $_->isa('AI::MXNet::NDArray'))
+            for @$vals;
+        return ([ ($keys) x @$vals ], [ map { $_->handle } @$vals ]);
+    }
+}
 
+# A wrapper for the user-defined handle.
+method _updater_wrapper($updater)
+{
+    return sub {
+        my($key, $lhs_handle, $rhs_handle) = @_;
+        # ctypes function
+        my $lhs = AI::MXNet::NDArray->new(handle => $lhs_handle);
+        my $rhs = AI::MXNet::NDArray->new(handle => $rhs_handle);
+        $updater->($key, $lhs, $rhs);
+    };
+}
 =head1 DESCRIPTION
 
-    Key value store interface of MXNet for parameter synchronization, over multiple devices.
+    A key-value store for synchronization of values, over multiple devices.
 =cut
 
 has 'handle' => (is => 'ro', isa => 'KVStoreHandle', required => 1);
 has '_updater' => (is => 'rw',  isa => 'AI::MXNet::Updater');
 has '_updater_func' => (is => 'rw', isa => 'CodeRef');
 
-sub DEMOLISH
+method DEMOLISH
```

Comment:
this should remain 'sub', method has to be defined as method(@args) and we do not use these args, so sub is preferred.